darts/models/torch_forecasting_model.py

```diff
@@ -536,6 +563,93 @@ def predict_from_dataset(self,
+        input_series_dataset
+            Optionally, one or several input `TimeSeries`, representing the history of the target series' whose
+            future is to be predicted. If specified, the method returns the forecasts of these
+            series. Otherwise, the method returns the forecast of the (single) training series.
+
+        Returns
+        -------
+        torch.Tensor of shape `(len(input_series_dataset), self.input_chunk_length, target.width + covariates.width)`
+            Initial input to model which contains the input target sereis and the matching past covariates, 
+            if available
+        Optional[torch.Tensor] of shape `(len(input_series_dataset), n, covariates.width)`
+            Covariates required to predict horizons beyond `self.output_chunk_length`, or `None` if the model
+            does require future covariates either because it does not require covariates at all or because
+            `n <= self.output_chunk_length`.
+    """
+
+        in_past_arr = []
+        cov_future_arr = []
+
+        for target_series, covariate_series in input_series_dataset:
+            raise_if_not(len(target_series) >= self.input_chunk_length,
+                            'All input series must have length >= `input_chunk_length` ({}).'.format(
+                self.input_chunk_length))
+
+            # TODO: here we could be smart and handle cases where target and covariates do not have same time axis.
+            # TODO: e.g. by taking their latest common timestamp.
+
+            in_past = target_series.values(copy=False)[-self.input_chunk_length:]
+            in_past = torch.from_numpy(in_past).float().to(self.device)
+
+            if covariate_series is not None:
+
+                # get first timestamp that lies in the future of target series
+                first_pred_time = target_series.end_time() + target_series.freq()
+                
+                # check whether future covariates are available and separate them if they are
+                if covariate_series.end_time() >= first_pred_time:
+                    cov_past = covariate_series.drop_after(first_pred_time)
+                else:
+                    cov_past = covariate_series
+
+                # keep only the last covariates that the model can use as input
+                cov_past = cov_past.values(copy=False)[-self.input_chunk_length:]
+
+                # create torch tensor
+                cov_past = torch.from_numpy(cov_past).float().to(self.device)
+
+                # combine target series with covariates
+                in_past = torch.cat([in_past, cov_past], dim=1)
+
+                # handle future covariates
+                if n > self.output_chunk_length:
+                    # drop past covariates (already used)
+                    cov_future = covariate_series.drop_before(first_pred_time - covariate_series.freq()).values(copy=False)
```

Comment:
Could we maybe do this dropping logic (here and in line 613) using integer indices of the time series, rather than using timestamps and freq? It will generalise better to new kinds of TimeSeries that may be indexed by RangeIndex instead of DatetimeIndex (that I'm working on).

Reply:
Right now I am using the last timestamp of the target series to establish when 'the future starts'. This way the covariates can also be longer into the past and it should still work. With integers I'm not sure I could replicate this behavior. Do you think that kind of robustness might not be necessary? I guess it wouldn't be hard to change, we just have to add the requirement that target series and covariates must all have the same starting point. I could add that requirement

Reply:
Ah I see. It's very nice to do as you say, actually (using timestamps to infer where the future starts). 
However this won't work for all series sometime soon; I'm working on a change of TimeSeries which will make them indexed either using a DatetimeIndex (like now) or a RangeIndex (for cases where users only have sequential data without timestamps). So for this logic here, what we should probably do is check if the series uses a DatetimeIndex, and if it doesn't then we should require the series to have the same starting point (in order to avoid relying on the timestamp). Makes sense?

Reply:
Makes sense to me. What do you think about doing this in a separate PR though?

Reply:
Yes, agree